### DISCUSSION TITLE
Allow shallow rounting on `pathname` - leading to Router bug

### DISCUSSION THREAD

**Discussion by Alecell** [vercel/next.js#23950] [Ideas]
# Feature request
It would be awesome if Next support `shallow` change URL without reload the page with `pathname`. We have so many cases that is useful and user-friendly change URL without redirect, only changing the current `pathname` as the  `history.pushState` does.

## Is your feature request related to a problem? Please describe.
Currently we can use the `pushState` method to change URL without reload, but it's followed with a bug that `asPath` has some strange behavior that sometimes it get the current URL correctly, but sometimes don't, it get the previous URL.

Currently I have a multiple list of, lest say, developers, each lists represent a list of developers main language. So for example we have 3 lists a list of Python developers, Node developers and C. The URL of this page is `/developers/[language]`, we start at `/developers/python`, but when we swipe left (thinking on mobile experience) I need the URL to change from `/developers/python` to `/developers/node`, no need of reload the page that hurts the user experience.

So far, the `pushState` handle the case, but if we have a click on the developer that leads to a URL like `/developers/[language]/[id]`, so lets suppose we start at `/developers/python`, swipe and go to `/developers/node` and click on a developer, at this point I'm facing a problem a problem that seems the next `Router` don't update its routes when `pushState` is used, so the `asPath` has the old URL `/developers/python` instead of `/developers/node`.

## Describe the solution you'd like
I know, by the way I believe, that is a expected error, as I'm changing the URL without the `Router` object so it makes sense to it don't update, but I'm using the `pushState` since Next don't handle `shallow` routing with `pathname`, only with `query params`.

It would be awesome if Next has the feature to allow `shallow` routing with `pathname` since there are use cases that makes much sense to change URL without reload the page. 

## Describe alternatives you've considered
The alternative is just the `pushState`, but it isn't much useful since it leads to a bug forcing to use the `window.location` and replacing the location when have url params.

## Additional context
I did not know if have a good reason to Next don't handle `shallow` routing with `pathname` but since I posted a question about this [here](https://github.com/vercel/next.js/discussions/15199) and I didn't get an answer, I beleived that isn't.

**Comment by kevinsimper** [comment]
I also tried using first `router.replace()` but that made the page think is has to reload.

As @Alecell I tried using `history.pushState`, but that breaks the navigation back in Next.js. When you go back, you don't go to the previous page, you stay on the same but the URL change.

**Comment by ijjk** [comment]
Hi, this sounds related to https://github.com/vercel/next.js/issues/16028 which was fixed in https://github.com/vercel/next.js/pull/16477 and is available on the latest canary of Next.js `v9.5.3-canary.21`. Can you upgrade and see if the behavior is corrected for you?

If I'm misunderstanding and you are instead requesting to be able to do shallow routing when navigating to a different page e.g. from `pages/blog/[slug].js` to `pages/different/[slug].js` then this isn't supported because these pages can have different data requirements and shallow routing skips calling `getStaticProps`/`getServerSideProps` for the page.
